### PR TITLE
Avoid zeroing large stack buffers in stdio on Windows

### PR DESCRIPTION
**Pull request by thomcc**
Does what it says on the tin, using `[MaybeUninit<u16>; N]` instead of `[0u16; N]`. These buffers seem to be around 8kb, which is big enough that this is likely to be a very nice perf boost to stdio-heavy windows code.

r? @ChrisDenton

*(Note: this PR also has a commit that adds windows to CI, but as it mentions I'll revert that after it comes out green -- I can only do a check build on the machine I'm typing this on)*

### CODE REVIEW

**Comment by thomcc** [comment]
CI is green on windows (https://github.com/rust-lang/rust/runs/8089709575?check_suite_focus=true and https://github.com/rust-lang/rust/runs/8089709693?check_suite_focus=true), so I dropped the CI edit commit, and this should be reviewable.

File: library/std/src/sys/windows/stdio.rs [hunk 1]
```diff
@@ -169,13 +170,14 @@ fn write(
 }
 
 fn write_valid_utf8_to_console(handle: c::HANDLE, utf8: &str) -> io::Result<usize> {
-    let mut utf16 = [0u16; MAX_BUFFER_SIZE / 2];
+    let mut utf16: [MaybeUninit<u16>; MAX_BUFFER_SIZE / 2] = MaybeUninit::uninit_array();
```
Comment (ChrisDenton) [inline library/std/src/sys/windows/stdio.rs:173]: You might want to replace uses of `uninit_array`. See #101179.

**Review by ChrisDenton — COMMENTED** [review]
Looks good. Just a couple of things

File: library/std/src/sys/windows/stdio.rs (hunk 1 above)
Comment (ChrisDenton) [reply library/std/src/sys/windows/stdio.rs:173]: Oh! Just noticed you commented there with some disagreement. I'll defer to your preference here as I don't really have one.

File: library/std/src/sys/windows/stdio.rs [hunk 2]
```diff
@@ -250,27 +252,33 @@ impl io::Read for Stdin {
             return Ok(bytes_copied);
         } else if buf.len() - bytes_copied < 4 {
             // Not enough space to get a UTF-8 byte. We will use the incomplete UTF8.
-            let mut utf16_buf = [0u16; 1];
+            let mut utf16_buf = [MaybeUninit::new(1); 1];
```
[the remark this answers is not in the document]
Comment (thomcc) [reply library/std/src/sys/windows/stdio.rs:255]: Ah, I misread the old code.

File: library/std/src/sys/windows/stdio.rs (hunk 1 above)
Comment (thomcc) [reply library/std/src/sys/windows/stdio.rs:173]: Not a huge fan but fair enough.

Comment (thomcc) [reply library/std/src/sys/windows/stdio.rs:173]: Nah I'll change it. I don't feel that strongly and don't feel like causing issues for someone during rollup.

**Pull request merged**